File: app/src/main/java/com/booboot/vndbandroid/api/SocketPool.java
package com.booboot.vndbandroid.api;

import android.content.Context;

import com.booboot.vndbandroid.util.Callback;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import javax.net.ssl.SSLSocket;

/**
 * Created by od on 04/06/2016.
 */
public class SocketPool {
    public final static int MAX_SOCKETS = 5;
    public final static SSLSocket[] SOCKETS = new SSLSocket[MAX_SOCKETS];
    public final static ConcurrentMap<Integer, Integer> LOCKS = new ConcurrentHashMap<>();

    public static SSLSocket getSocket(Context context, int socketIndex, Callback errorCallback) {
        socketIndex %= SOCKETS.length;
        VNDBServer.login(context, socketIndex, errorCallback);
        return SOCKETS[socketIndex];
    }

    public static void setSocket(int socketIndex, SSLSocket socket) {
        SOCKETS[socketIndex % SOCKETS.length] = socket;
    }

    public static SSLSocket getSocket(int socketIndex) {
        return SOCKETS[socketIndex % SOCKETS.length];
    }

    public static Object getLock(Integer id) {
        id %= SOCKETS.length;
        LOCKS.putIfAbsent(id, id);
        return LOCKS.get(id);
    }
}

File: app/src/main/java/com/booboot/vndbandroid/api/SocketPool.kt
package com.booboot.vndbandroid.api

import com.booboot.vndbandroid.model.vndb.Options
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

import javax.net.ssl.SSLSocket

object SocketPool {
    const val MAX_SOCKETS = 10
    private val SOCKETS = arrayOfNulls<SSLSocket>(MAX_SOCKETS)
    private val LOCKS: ConcurrentMap<Int, Int> = ConcurrentHashMap()
    var throttleHandlingSocket = -1

    fun getSocket(server: VNDBServer, options: Options): SSLSocket? {
        server.login(options)
        return SOCKETS[options.socketIndex]
    }

    fun getSocket(socketIndex: Int): SSLSocket? = SOCKETS[socketIndex]

    fun setSocket(socketIndex: Int, socket: SSLSocket?) {
        SOCKETS[socketIndex] = socket
    }

    fun getLock(id: Int): Int {
        LOCKS.putIfAbsent(id, id)
        return LOCKS[id] ?: 0
    }
}